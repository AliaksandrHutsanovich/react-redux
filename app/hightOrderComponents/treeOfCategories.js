import React, { memo } from 'react';
import 'antd/dist/antd.css';
import { Tree } from 'antd';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { getCategories } from '../selectors/selectors';

const { TreeNode } = Tree;

const treeOfCategories = (TreeNodeTitle) => {
  const getTreeNodes = (categories) => categories.map((category) => {
    if (category.subCategories.length) {
      return (
        <TreeNode
          key={category.key}
          className="tree_node"
          title={<TreeNodeTitle path={category.key} title={category.title} />}
        >
          {getTreeNodes(category.subCategories)}
        </TreeNode>
      );
    }
    return (
      <TreeNode
        key={category.key}
        className="tree_node"
        title={(
          <TreeNodeTitle
            path={category.key}
            title={category.title}
          />
        )}
      />
    );
  });

  const TreeOfCategories = ({ onSelectCategory, categories }) => (
    <Tree
      className="draggable-tree"
      draggable
      onSelect={onSelectCategory}
    >
      {getTreeNodes(categories)}
    </Tree>
  );

  TreeOfCategories.defaultProps = {
    onSelectCategory: () => {},
  };

  TreeOfCategories.propTypes = {
    onSelectCategory: PropTypes.func,
    categories: PropTypes.arrayOf(PropTypes.object).isRequired,
  };

  const mapStateToProps = (state) => ({ categories: getCategories(state) });

  return connect(mapStateToProps)(memo(TreeOfCategories));
};

export default treeOfCategories;
