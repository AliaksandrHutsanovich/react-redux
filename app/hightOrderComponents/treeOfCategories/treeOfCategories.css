.draggableTree {
    margin-top: 20px !important;
    width: 90%;
    align-self: center;
    max-height: 85%;
    overflow: auto;
}