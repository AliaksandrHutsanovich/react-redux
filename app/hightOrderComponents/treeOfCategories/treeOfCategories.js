import React, { memo } from 'react';
import '!style-loader!css-loader!antd/dist/antd.css'; // eslint-disable-line
import { Tree } from 'antd';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { getCategories } from '../../selectors/selectors';

import styles from './treeOfCategories.css';

const { TreeNode } = Tree;

const treeOfCategories = (TreeNodeTitle) => {
  const getTreeNode = (category, getChildNodes) => (
    <TreeNode
      key={category.key}
      title={<TreeNodeTitle path={category.key} title={category.title} />}
    >
      {getChildNodes && getChildNodes(category.subCategories)}
    </TreeNode>
  );

  const getTreeNodes = (categories) => categories.map((category) => {
    if (category.subCategories.length) {
      return getTreeNode(category, getTreeNodes);
    }
    return getTreeNode(category);
  });

  const TreeOfCategories = ({ onSelectCategory, categories }) => (
    <Tree
      className={styles.draggableTree}
      draggable
      onSelect={onSelectCategory}
    >
      {getTreeNodes(categories)}
    </Tree>
  );

  TreeOfCategories.defaultProps = {
    onSelectCategory: () => {},
  };

  TreeOfCategories.propTypes = {
    onSelectCategory: PropTypes.func,
    categories: PropTypes.arrayOf(PropTypes.object).isRequired,
  };

  const mapStateToProps = (state) => ({ categories: getCategories(state) });

  return connect(mapStateToProps)(memo(TreeOfCategories));
};

export default treeOfCategories;
