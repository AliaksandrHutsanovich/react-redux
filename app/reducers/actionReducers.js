import {
    addCategory,
    addSubCategory,
    deleteCategory,
    editCategory,
    editTaskStatus,
    addTask,
    editTask,
    deleteTask,
    insertCategory,
    insertTask,
    editCategoryRevive,
    editTaskRevive
} from '../actions/actions';
import { handleActions } from 'redux-actions';
import immutable from 'immutable';
const Map = immutable.Map,
    List = immutable.List;
import { initialState } from './states/initialState';

export const actionReducers = handleActions({

    [addCategory]: {
        next(state, { payload }) {
            return state
                .update("categories",
                (categories) => categories.push(Map({ title: payload.title, subCategories: List([]), tasks: List([]) })));
        }
    },

    [addSubCategory]: {
        next(state, { payload }) {
            return state
                .updateIn(payload.path,
                subCategories => subCategories.push(Map({ title: payload.title, subCategories: List([]), tasks: List([]) })));
        }
    },

    [deleteCategory]: {
        next(state, { payload }) {
            let fullPath = Object.assign([], payload.path);
            fullPath.push(payload.pathParam);
            return state.deleteIn(fullPath);
        }
    },

    [editCategory]: {
        next(state, { payload }) {
            let fullPath = Object.assign([], payload.path);
            fullPath.push(payload.pathParam);
            return state
                .updateIn(fullPath,
                category => {
                    return category.set("title", payload.title);
                });
        }
    },

    [editTaskStatus]: {
        next(state, { payload }) {
            let fullPath = Object.assign([], payload.path);
            fullPath.push(payload.pathParam);
            return state
                .updateIn(fullPath,
                task => {
                    return task.set("isFinished", payload.value);
                });
        }
    },

    [addTask]: {
        next(state, { payload }) {
            return state
                .updateIn(payload.path,
                tasks => tasks.push(Map({ title: payload.title, Description: "", isFinished: false })));
        }
    },

    [editTask]: {
        next(state, { payload }) {
            if (payload.newPath) {
                if (payload.newPathParam) {
                    state = state
                        .updateIn(payload.newPath,
                        tasks => tasks.insert(payload.newPathParam, Map({ title: payload.title, Description: payload.description, isFinished: payload.isFinished })));
                }
                else {
                    state = state
                        .updateIn(payload.newPath,
                        tasks => tasks.push(Map({ title: payload.title, Description: payload.description, isFinished: payload.isFinished })));
                }
                if (payload.oldPath) {
                    let fullOldPath = Object.assign([], payload.oldPath);
                    fullOldPath.push(payload.oldPathParam);
                    return state.deleteIn(fullOldPath);
                }
                else {
                    return state;
                }
            }
            else {
                let fullOldPath = Object.assign([], payload.oldPath);
                fullOldPath.push(payload.oldPathParam);
                return state
                    .updateIn(fullOldPath,
                    task => task.set('title', payload.title).set('Description', payload.description).set('isFinished', payload.isFinished));
            }
        }
    },

    [deleteTask]: {
        next(state, { payload }) {
            let fullPath = Object.assign([], payload.path);
            fullPath.push(payload.pathParam);
            return state.deleteIn(fullPath);
        }
    },

    [insertCategory]: {
        next(state, { payload }) {
            return state.updateIn(payload.path, categories => categories.insert(payload.pathParam, Map({ title: payload.title, subCategories: payload.subCategories, tasks: payload.tasks })));
        }
    },

    [insertTask]: {
        next(state, { payload }) {
            return state.updateIn(payload.path, tasks => tasks.insert(payload.pathParam, Map({ title: payload.title, Description: payload.description, isFinished: payload.isFinished })));
        }
    },

    [editCategoryRevive]: {
        next(state, { payload }) {
            let fullPath = Object.assign([], payload.path);
            fullPath.push(payload.pathParam);
            return state
                .updateIn(fullPath,
                category => {
                    return category.set("title", payload.titlePrimary);
                });
        }
    },

    [editTaskRevive]: {
        next(state, { payload }) {
            if (payload.newPath) {
                let fullNewPath = Object.assign([], payload.newPath);
                fullNewPath.push(payload.newPathParam);

                return state
                    .updateIn(payload.oldPath,
                    tasks => tasks.insert(payload.oldPathParam, Map({ title: payload.titlePrimary, Description: payload.descriptionPrimary, isFinished: payload.isFinishedValuePrimary })))
                    .deleteIn(fullNewPath);
            }
            else {
                let fullOldPath = Object.assign([], payload.oldPath);
                fullOldPath.push(payload.oldPathParam);
                return state.updateIn(fullOldPath, task => task.set('title', payload.titlePrimary).set('Description', payload.descriptionPrimary).set('isFinished', payload.isFinishedValuePrimary));
            }
        }
    }

}, Map().merge(initialState));