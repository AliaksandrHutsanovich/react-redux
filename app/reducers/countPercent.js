import { countPercentage, incrementInDone, decrementInDone, incrementInAll, decrementInAll } from '../actions/actions';
import { handleActions } from 'redux-actions';
import immutable from 'immutable';
const Map = immutable.Map;

export const countPer—Åent = handleActions({

    [countPercentage]: {
        next(state, { payload }) {
            return state.set('done', payload.done).set('all', payload.all);
        }
    },

    [incrementInDone]: {
        next(state, { payload }) {
            let done = state.get('done') + 1;
            return state.set('done', done);
        }
    },

    [decrementInDone]: {
        next(state, { payload }) {
            let done = state.get('done') - 1;
            return state.set('done', done);
        }
    },

    [incrementInAll]: {
        next(state, { payload }) {
            let all = state.get('all') + 1;
            return state.set('all', all);
        }
    },

    [decrementInAll]: {
        next(state, { payload }) {
            let all = state.get('all') - 1;
            return state.set('all', all);
        }
    }

}, Map({ done: 0, all: 0 }));

