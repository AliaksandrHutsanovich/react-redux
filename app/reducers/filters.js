import { Map } from 'immutable';
import { handleActions } from 'redux-actions';
import {
  showDoneTasks,
  searchTasks,
} from '../actions';

const filters = handleActions({

  [showDoneTasks]: {
    next(state, { payload }) {
      return state.set('showDone', payload);
    },
  },

  [searchTasks]: {
    next(state, { payload }) {
      return state.set('searchKey', payload);
    },
  },

}, Map({ showDone: true, searchKey: '' }));

export default filters;
