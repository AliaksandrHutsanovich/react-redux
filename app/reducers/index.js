export { default } from './reducer';
export { initialState } from './states/initialState';
