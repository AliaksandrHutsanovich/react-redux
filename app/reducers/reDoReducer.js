import { Map, List } from 'immutable';
import { handleActions } from 'redux-actions';
import { addToReDo, deleteFromReDo, clearReDo } from '../actions';

const reDoReducer = handleActions({

  [addToReDo]: {
    next(state, { payload }) {
      return state
        .update('redoOperations', (operations) => operations.push(Map(payload)));
    },
  },

  [deleteFromReDo]: {
    next(state) {
      return state.update('redoOperations', (operations) => operations.delete(state.get('redoOperations').toArray().length - 1));
    },
  },

  [clearReDo]: {
    next(state) {
      return state.set('redoOperations', List([]));
    },
  },

}, Map({ redoOperations: List([]) }));

export default reDoReducer;
