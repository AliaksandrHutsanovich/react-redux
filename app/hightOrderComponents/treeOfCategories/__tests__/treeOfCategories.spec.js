import React from 'react';
import { shallow, mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Map } from 'immutable';
import treeOfCategories from '../treeOfCategories';
import { initialState as store } from '../../../reducers';
import CategoryItem from '../../../layout/layoutContent/tasks/categoryItem';

const mockStore = configureStore();
const initialState = Map().merge(store);
const TreeOfCategories = treeOfCategories(CategoryItem);

describe('Unit tests for tree of categories', () => {
  const dataStore = mockStore({ actionReducers: initialState });
  const Component = shallow(<TreeOfCategories store={dataStore} />);

  it('full render test', () => {
    expect(Component).toMatchSnapshot();
    mount(<TreeOfCategories store={dataStore} />);
  });
});
