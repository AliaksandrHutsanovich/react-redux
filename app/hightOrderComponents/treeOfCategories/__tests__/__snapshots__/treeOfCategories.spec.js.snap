// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for tree of categories full render test 1`] = `
<Connect(Component)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <TreeOfCategories
    categories={
      Array [
        Object {
          "key": "categories-0",
          "subCategories": Array [
            Object {
              "key": "categories-0-subCategories-0",
              "subCategories": Array [],
              "tasks": Immutable.List [
                Immutable.Map {
                  "title": "integral",
                  "Description": "find",
                  "isFinished": false,
                },
                Immutable.Map {
                  "title": "different",
                  "Description": "find",
                  "isFinished": false,
                },
              ],
              "title": "highMathematic",
            },
            Object {
              "key": "categories-0-subCategories-1",
              "subCategories": Array [],
              "tasks": Immutable.List [
                Immutable.Map {
                  "title": "table",
                  "Description": "constit",
                  "isFinished": true,
                },
                Immutable.Map {
                  "title": "multi",
                  "Description": "count",
                  "isFinished": false,
                },
              ],
              "title": "lowMathematic",
            },
          ],
          "tasks": Immutable.List [
            Immutable.Map {
              "title": "graph",
              "Description": "draw",
              "isFinished": true,
            },
            Immutable.Map {
              "title": "counting",
              "Description": "count",
              "isFinished": true,
            },
          ],
          "title": "Mathematic",
        },
        Object {
          "key": "categories-1",
          "subCategories": Array [
            Object {
              "key": "categories-1-subCategories-0",
              "subCategories": Array [
                Object {
                  "key": "categories-1-subCategories-0-subCategories-0",
                  "subCategories": Array [],
                  "tasks": Immutable.List [
                    Immutable.Map {
                      "title": "peoples",
                      "Description": "study",
                      "isFinished": true,
                    },
                    Immutable.Map {
                      "title": "capitals",
                      "Description": "study",
                      "isFinished": false,
                    },
                  ],
                  "title": "smallHighGeography",
                },
              ],
              "tasks": Immutable.List [
                Immutable.Map {
                  "title": "polit",
                  "Description": "study",
                  "isFinished": false,
                },
                Immutable.Map {
                  "title": "mountain",
                  "Description": "study",
                  "isFinished": false,
                },
              ],
              "title": "highGeography",
            },
          ],
          "tasks": Immutable.List [
            Immutable.Map {
              "title": "map",
              "Description": "study",
              "isFinished": false,
            },
            Immutable.Map {
              "title": "drawing",
              "Description": "draw",
              "isFinished": false,
            },
          ],
          "title": "Geography",
        },
      ]
    }
    dispatch={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <Tree
      checkable={false}
      className="draggableTree"
      draggable={true}
      onSelect={[Function]}
      openAnimation={
        Object {
          "appear": null,
          "enter": [Function],
          "leave": [Function],
        }
      }
      showIcon={false}
    >
      <Tree
        autoExpandParent={false}
        checkStrictly={false}
        checkable={false}
        className="ant-tree-icon-hide draggableTree"
        defaultCheckedKeys={Array []}
        defaultExpandAll={false}
        defaultExpandParent={true}
        defaultExpandedKeys={Array []}
        defaultSelectedKeys={Array []}
        disabled={false}
        draggable={true}
        multiple={false}
        onSelect={[Function]}
        openAnimation={
          Object {
            "appear": null,
            "enter": [Function],
            "leave": [Function],
          }
        }
        prefixCls="ant-tree"
        selectable={true}
        showIcon={false}
        showLine={false}
        switcherIcon={[Function]}
      >
        <ul
          className="ant-tree ant-tree-icon-hide draggableTree"
          role="tree"
          unselectable="on"
        >
          <TreeNode
            checked={false}
            dragOver={false}
            dragOverGapBottom={false}
            dragOverGapTop={false}
            eventKey="categories-0"
            expanded={false}
            halfChecked={false}
            key="categories-0"
            loaded={false}
            loading={false}
            pos="0-0"
            selected={false}
            title={
              <Memo(CategoryItem)
                path="categories-0"
                title="Mathematic"
              />
            }
          >
            <li
              className="ant-tree-treenode-switcher-close"
              onDragEnd={[Function]}
              onDragEnter={[Function]}
              onDragLeave={[Function]}
              onDragOver={[Function]}
              onDrop={[Function]}
              role="treeitem"
            >
              <span
                className="ant-tree-switcher ant-tree-switcher_close"
                onClick={[Function]}
              >
                <switcherIcon
                  checked={false}
                  dragOver={false}
                  dragOverGapBottom={false}
                  dragOverGapTop={false}
                  eventKey="categories-0"
                  expanded={false}
                  halfChecked={false}
                  isLeaf={false}
                  loaded={false}
                  loading={false}
                  pos="0-0"
                  selected={false}
                  title={
                    <Memo(CategoryItem)
                      path="categories-0"
                      title="Mathematic"
                    />
                  }
                >
                  <Icon
                    className="ant-tree-switcher-icon"
                    theme="filled"
                    type="caret-down"
                  >
                    <LocaleReceiver
                      componentName="Icon"
                    >
                      <i
                        aria-label="icon: caret-down"
                        className="anticon anticon-caret-down ant-tree-switcher-icon"
                      >
                        <IconReact
                          className=""
                          type="caret-down-fill"
                        >
                          <svg
                            aria-hidden="true"
                            className=""
                            data-icon="caret-down"
                            fill="currentColor"
                            height="1em"
                            key="svg-caret-down"
                            viewBox="0 0 1024 1024"
                            width="1em"
                          >
                            <path
                              d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                              key="svg-caret-down-svg-0"
                            />
                          </svg>
                        </IconReact>
                      </i>
                    </LocaleReceiver>
                  </Icon>
                </switcherIcon>
              </span>
              <span
                aria-grabbed={true}
                className="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-close draggable"
                draggable={true}
                onClick={[Function]}
                onContextMenu={[Function]}
                onDoubleClick={[Function]}
                onDragStart={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                title=""
              >
                <span
                  className="ant-tree-title"
                >
                  <Memo(CategoryItem)
                    path="categories-0"
                    title="Mathematic"
                  >
                    <p
                      className="itemText"
                    >
                      Mathematic
                    </p>
                  </Memo(CategoryItem)>
                </span>
              </span>
              <Animate
                animation={
                  Object {
                    "appear": null,
                    "enter": [Function],
                    "leave": [Function],
                  }
                }
                component=""
                componentProps={Object {}}
                showProp="data-expanded"
                transitionAppear={false}
                transitionEnter={true}
                transitionLeave={true}
              />
            </li>
          </TreeNode>
          <TreeNode
            checked={false}
            dragOver={false}
            dragOverGapBottom={false}
            dragOverGapTop={false}
            eventKey="categories-1"
            expanded={false}
            halfChecked={false}
            key="categories-1"
            loaded={false}
            loading={false}
            pos="0-1"
            selected={false}
            title={
              <Memo(CategoryItem)
                path="categories-1"
                title="Geography"
              />
            }
          >
            <li
              className="ant-tree-treenode-switcher-close"
              onDragEnd={[Function]}
              onDragEnter={[Function]}
              onDragLeave={[Function]}
              onDragOver={[Function]}
              onDrop={[Function]}
              role="treeitem"
            >
              <span
                className="ant-tree-switcher ant-tree-switcher_close"
                onClick={[Function]}
              >
                <switcherIcon
                  checked={false}
                  dragOver={false}
                  dragOverGapBottom={false}
                  dragOverGapTop={false}
                  eventKey="categories-1"
                  expanded={false}
                  halfChecked={false}
                  isLeaf={false}
                  loaded={false}
                  loading={false}
                  pos="0-1"
                  selected={false}
                  title={
                    <Memo(CategoryItem)
                      path="categories-1"
                      title="Geography"
                    />
                  }
                >
                  <Icon
                    className="ant-tree-switcher-icon"
                    theme="filled"
                    type="caret-down"
                  >
                    <LocaleReceiver
                      componentName="Icon"
                    >
                      <i
                        aria-label="icon: caret-down"
                        className="anticon anticon-caret-down ant-tree-switcher-icon"
                      >
                        <IconReact
                          className=""
                          type="caret-down-fill"
                        >
                          <svg
                            aria-hidden="true"
                            className=""
                            data-icon="caret-down"
                            fill="currentColor"
                            height="1em"
                            key="svg-caret-down"
                            viewBox="0 0 1024 1024"
                            width="1em"
                          >
                            <path
                              d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                              key="svg-caret-down-svg-0"
                            />
                          </svg>
                        </IconReact>
                      </i>
                    </LocaleReceiver>
                  </Icon>
                </switcherIcon>
              </span>
              <span
                aria-grabbed={true}
                className="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-close draggable"
                draggable={true}
                onClick={[Function]}
                onContextMenu={[Function]}
                onDoubleClick={[Function]}
                onDragStart={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                title=""
              >
                <span
                  className="ant-tree-title"
                >
                  <Memo(CategoryItem)
                    path="categories-1"
                    title="Geography"
                  >
                    <p
                      className="itemText"
                    >
                      Geography
                    </p>
                  </Memo(CategoryItem)>
                </span>
              </span>
              <Animate
                animation={
                  Object {
                    "appear": null,
                    "enter": [Function],
                    "leave": [Function],
                  }
                }
                component=""
                componentProps={Object {}}
                showProp="data-expanded"
                transitionAppear={false}
                transitionEnter={true}
                transitionLeave={true}
              />
            </li>
          </TreeNode>
        </ul>
      </Tree>
    </Tree>
  </TreeOfCategories>
</Connect(Component)>
`;
